### DISCUSSION TITLE
android-tools with Termux-usb

### DISCUSSION THREAD

**Discussion by ghost** [termux/termux-packages#10298] [Ideas]
~~[patch here](https://gist.github.com/ookiineko/b15412b6119e3560e8cfd20e0c807ba4)~~

https://user-images.githubusercontent.com/103345643/164944224-e9f35a32-7c2d-4f5d-8489-ce7a38f3a323.mp4

**Comment by ghost** [comment]
this is just a PoC xdd will send PR when ready

**Comment by ghost** [comment]
fastbootd works
flashing twrp & gsi works

**Comment by ghost** [comment]
fastboot getvar works
fastboot reboot works

**Comment by ghost** [comment]
i restored fastboot devices output xddd

<img src="https://user-images.githubusercontent.com/103345643/165229942-76019106-b79f-457d-b978-d112c9585caf.png" width=350/>

**Reply by ghost** [reply]
will send PR later xdd

**Comment by anshiq** [comment]
Can you help please tell what is this fastboot can i root my mobile using this

**Reply by ghost** [reply]
ig you can, but first you have to unlock your bootloader

**Reply by ghost** [reply]
after that grab the boot.img from stock rom and copy it into your device's internal storage,
and patch it using Magisk Manager (no root required)

then you can just simply flash it with `fastboot flash boot <your magisk patched boot>.img`
you need at least two devices (phone or computer) to do this, no root is required

**Reply by ghost** [reply]
you may also want to flash an empty vbmeta.img to vbmeta* (vbmeta, vbmeta_system, vbmeta_vendor, etc.), in order to bypass avb 2.0

**Reply by ghost** [reply]
also, you can just flash a TWRP or other recovery project, if there's one for your device
and then you can install Magisk on it, it's probably easier

**Reply by ghost** [reply]
I can't tell you the specific steps, depending on your device, the way may be different

**Reply by anshiq** [reply]
thankyou, @ookiineko  that enough bro now I have an Idea

**Reply by ghost** [reply]
if you have Telegram i can try to help as possible, email me your username or sth

**Reply by ghost** [reply]
> thankyou, @ookiineko that enough bro now I have an Idea

ok

**Reply by anshiq** [reply]
> if you have Telegram i can try to help as possible, email me your username or sth

ok I am mailing you my telegram username.

**Reply by Buffalo-716** [reply]
> ig you can, but first you have to unlock your bootloader

Im interested in learning how to do this also they say my ZTE blade Vantage 2 is hard to do anything with. Any ideals? You seem very smart about this work around root access.

**Comment by anshiq** [comment]
Hlo Bro can you help me, I want to compile vs code for termux. I am absolute beginner.

**Reply by ghost** [reply]
you don't have to do that ig, proot/chroot containers are recommended, and you can install software like vscode in it xdd

**Reply by anshiq** [reply]
Bro I have  low specs smart phone. termux gets crashes when i install desktop envirment in any linux distro thats why :)

**Reply by ghost** [reply]
> Bro I have low specs smart phone. termux gets crashes when i install desktop envirment in any linux distro thats why :)

maybe u could try sth like dwm or i3wm? which are much lighter than a full desktop env
and still allow u to run GUI apps

**Reply by ghost** [reply]
~~u may want xserver xsdl as well to do that~~ try termux:x11

**Reply by ghost** [reply]
and also, code-server does not require a desktop env to use, u can run it on any standard linux env (like proot/chroot of any distro), and access it from your browser

there is also an app optimized for Android that can run the code-server directly: https://github.com/vhqtvn/VHEditor-Android

**Comment by RohitVerma882** [comment]


**Reply by ghost** [reply]
i am sorry but i am not working on it

**Reply by RohitVerma882** [reply]
Then help me bro.

**Reply by ghost** [reply]
> I'm waiting for adb support bro with no root.

will try

**Reply by RohitVerma882** [reply]
Thank you!

**Reply by ghost** [reply]
> What i want to Modify on adb sources

rm deps on /sys/bus/usb/devices and /dev/bus/usb and create device directly from TERMUX_USB_FD

**Reply by RohitVerma882** [reply]
> > What i want to Modify on adb sources
> 
> rm deps on /sys/bus/usb/devices and /dev/bus/usb and create device directly from TERMUX_USB_FD

I think it's not enough for me!

**Reply by ghost** [reply]
> I'm able to compile adb

why not

**Reply by ghost** [reply]
> Can you give us the patch files for adb??

not yet

**Reply by ghost** [reply]
i am sry

**Reply by RohitVerma882** [reply]
In adb source has ```usb_libusb.cpp``` and ```usb_linux.cpp``` Which of the following would be correct to edit.

**Reply by RohitVerma882** [reply]
```usb_libusb.cpp``` uses libusb

**Reply by RohitVerma882** [reply]
> In adb source has `usb_libusb.cpp` and `usb_linux.cpp` Which of the following would be correct to edit.

?

**Reply by ghost** [reply]
> > In adb source has `usb_libusb.cpp` and `usb_linux.cpp` Which of the following would be correct to edit.
> 
> ?

both

**Reply by ghost** [reply]
> In adb source has `usb_libusb.cpp` and `usb_linux.cpp` Which of the following would be correct to edit.

![image](https://user-images.githubusercontent.com/103345643/168214630-8d629eb4-f4dd-4e4a-8e58-6ca4238dbb3c.png)

when the ADB_LIBUSB environment variable is set to 1, it will use libusb::usb_init()
otherwise it will use usb_init() in usb_linux.cpp instead of libusb i guess

![image](https://user-images.githubusercontent.com/103345643/168214652-c5be4a00-322c-4513-a558-7cfb611a7ea3.png)

**Reply by RohitVerma882** [reply]
> > In adb source has `usb_libusb.cpp` and `usb_linux.cpp` Which of the following would be correct to edit.
> 
> ![image](https://user-images.githubusercontent.com/103345643/168214630-8d629eb4-f4dd-4e4a-8e58-6ca4238dbb3c.png)
> 
> when the ADB_LIBUSB environment variable is set to 1, it will use libusb::usb_init() otherwise it will use usb_init() in usb_linux.cpp instead of libusb i guess
> 
> ![image](https://user-images.githubusercontent.com/103345643/168214652-c5be4a00-322c-4513-a558-7cfb611a7ea3.png)

I'm modified ```linux_usb.cpp```, but it doesn't working!
Debugging dialog not showing

**Reply by RohitVerma882** [reply]
![Screenshot_2022-05-17-20-38-07-71_84d3000e3f4017145260f7618db1d683](https://user-images.githubusercontent.com/56387351/168845161-11f78a29-e825-4dad-8caf-94be46b37e91.jpg)
Now solved and working with root and nonroot.

**Reply by ghost** [reply]
nice work! send pr when

**Comment by Sushilku88** [comment]
Is it necessary to be root user to send fastboot commands?

**Comment by nohajc** [comment]
So, I noticed the fastboot patch was already merged but there's no corresponding adb patch. I wrote an alternative solution which lets you use unmodified adb from the `android-tools` package without root:
https://github.com/nohajc/termux-adb

**Reply by Sushilku88** [reply]
It means i send fastboot commands without root.

**Reply by nohajc** [reply]
Yes, it is possible. There just hasn't been any convenient way to do it yet.

**Reply by Sushilku88** [reply]
Thanks bro

**Reply by RohitVerma882** [reply]
> So, I noticed the fastboot patch was already merged but there's no corresponding adb patch. I wrote an alternative solution which lets you use unmodified adb from the `android-tools` package without root: https://github.com/nohajc/termux-adb

Not working in Android 12!
![Screenshot_2022-11-06-07-07-28-02_84d3000e3f4017145260f7618db1d683](https://user-images.githubusercontent.com/56387351/200150049-86624ce2-b9cf-4047-a5cd-996ce2f85b81.jpg)

**Reply by ghost** [reply]
wow it's so cool. I forgot we can use hooks. so this will be separated into a new package? looking forward to seeing you add it to termux-packages?

**Reply by ghost** [reply]
This is obviously a much better way to do it without patching the original source code, and it may not be limited to just adb and fastboot. i think it can be used for any program that requires usb access

**Reply by ghost** [reply]
> wow it's so cool. I forgot we can use hooks. so this will be separated into a new package? looking forward to seeing you add it to termux-packages?

just send a PR and revert my patches xdd

**Reply by Sushilku88** [reply]
That fastboot commands will also run in this package ?

**Reply by nohajc** [reply]
> This is obviously a much better way to do it without patching the original source code, and it may not be limited to just adb and fastboot. i think it can be used for any program that requires usb access

It's much better in theory. We already have at least one device where it doesn't work. Function hooking is tricky because you sometimes make assumptions which might not be true for every device.

So it will take some thorough testing to see if it's really usable. For me it was mostly a fun experiment. :D

**Reply by nohajc** [reply]
> That fastboot commands will also run in this package ?

Fastboot was already patched to support file descriptors passed from termux-usb. I just wrote a wrapper (termux-fastboot), so you can try that.

**Reply by nohajc** [reply]
> Not working in Android 12! ![Screenshot_2022-11-06-07-07-28-02_84d3000e3f4017145260f7618db1d683](https://user-images.githubusercontent.com/56387351/200150049-86624ce2-b9cf-4047-a5cd-996ce2f85b81.jpg)

Thank you for reporting the issue! When I have more information, I'll look into it.

**Comment by fraterakiba** [comment]
Can anyone help? 

$ pkg update
Checking availability of current mirror: ok
Hit:1 https://termux.mentality.rip/termux-packages-24 stable InRelease
Ign:2 https://dl.bintray.com/grimler/game-packages-24 games InRelease
Ign:3 https://dl.bintray.com/grimler/science-packages-24 science InRelease
Err:4 https://dl.bintray.com/grimler/game-packages-24 games Release
  502  Bad Gateway
Err:5 https://dl.bintray.com/grimler/science-packages-24 science Release
  502  Bad Gateway
Reading package lists... Done
E: The repository 'https://dl.bintray.com/grimler/game-packages-24 games Release' does not have a Release file.
N: Metadata integrity can't be verified, repository is disabled now.
N: Possible cause: repository is under maintenance or down (wrong sources.list URL?).
E: The repository 'https://dl.bintray.com/grimler/science-packages-24 science Release' does not have a Release file.
N: Metadata integrity can't be verified, repository is disabled now.
N: Possible cause: repository is under maintenance or down (wrong sources.list

**Reply by nohajc** [reply]
Try to set a different package mirror by running ` termux-change-repo`

**Reply by TomJo2000** [reply]
As of May 2021 Bintray has discontinued their hosting services, as such the Bintray repositories are defunct.

It is likely that you are running a very out of date version of Termux, such as the Google Playstore version, which has not been updated since release v0.101 due to Google Play now requires the target API to be ≥28 which causes some issues on Android ≥10.
Please consider switching to the [F-droid](https://f-droid.org/packages/com.termux/) or [GitHub](https://github.com/termux/termux-app/releases/latest/) release.
> Additional information (paraphrased from the Playstore informational tag from the Termux discord)
> **To avoid problems while switching:**
> In Settings > Applications, Find all applications on your phone named Termux:
> Such as.
> - Termux:API
> - Termux:Styling
> - Termux:Widget
> - Termux:Task
> - Termux:Float
> - Termux:Boot
> Then uninstall all of them (yes, including paid ones)
> If you need help with backing up your current files:
> https://wiki.termux.com/wiki/Backing_up_Termux

**Reply by Sushilku88** [reply]
Hi nohajc can you modify xioami adb for termux to sideloading

**Reply by nohajc** [reply]
> Hi nohajc can you modify xioami adb for termux to sideloading

Hi, do you mean this? https://www.xiaomitool.com/adb
It looks like there is no source code available. Only compiled `exe` for Windows. Unless there's also an ARM version, there's practically no way to run this in Termux.

**Comment by Sushilku88** [comment]
Yes bro

**Reply by nohajc** [reply]
Yeah... So, in principle, this could work but the author of xiaomi adb mod would have to release the source code or make an ARM build compatible with Termux.

**Reply by Sushilku88** [reply]
What about this tool https://github.com/Szaki/XiaomiADBFastbootTools

**Reply by Sushilku88** [reply]
This is the tool https://github.com/francescotescari/XiaoMiToolV2 which has xioami adb source code

**Reply by Sushilku88** [reply]
hi nohajc will you create xiomiadb package for termux?

**Reply by nohajc** [reply]
I don't know what I'm looking at here. Seems like some java app for PC. Not a patched version of adb. What is your use case exactly?

**Reply by Sushilku88** [reply]
I want to flash stock miui rom using stock recovery sideloadind by termux

**Reply by Sushilku88** [reply]
![Screenshot_7-2](https://user-images.githubusercontent.com/88477609/201895193-d8bd06f4-dd8a-436a-b0e5-822882ecfca1.png)

**Reply by nohajc** [reply]
That's nice but I can't tell yet if this is something in the scope of my original project.

**Reply by Sushilku88** [reply]
Ok thnks bro! but you must try it.

**Reply by nohajc** [reply]
I might try but no promises. 😄

**Reply by Sushilku88** [reply]
Ok ❤️

**Comment by Sushilku88** [comment]
Any tool in termux for xioami adb sideloadind
